xla/service/gpu/fusions/transpose_mlir.cc

```diff
@@ -131,14 +206,16 @@ std::optional<IndexingMap> MlirTransposeFusion::ComputeThreadIdToOutputIndexing(
                           .getSubMap(std::vector<unsigned>{permutation_.begin(),
                                                            permutation_.end()});
   auto thread_offset = GetThreadOffsetsForTiling(tiling_, mlir_context);
+  auto bitcast_indexing =
+      GetPermuteBlockTileBitcastMap(tiling_, permutation_, mlir_context);
   auto permuted_tiled_shape =
       ShapeUtil::MakeShape(U8, Permute(tiling_.GetShape(), permutation_));
 
   auto map = ComposeIndexingMaps(
       GetIndexingMapForTiling(
-          block_offset, thread_offset, tiling_.GetNumThreadsPerBlock(),
-          tiling_.GetNumBlocks(), tiling_.GetThreadTileSize(),
-          permuted_tiled_shape.dimensions()),
+          block_offset, bitcast_indexing.GetAffineMap().compose(thread_offset),
```

Comment:
This doesn't look quite right too me. Only the blocks should be permuted, the read/write pattern inside a block should be the same. Or am I missing something?

Reply:
We use [128, 256, 512]->[128, 512, 256] as example. Due to we add vectorized dim as the last dim of shape, so the input and output shape of transpose are [128, 256, 256, 2] - > [128, 256, 2, 256]. If block is permuted, the range of block offset are [[0, 127], [0, 255], [0, 1], [0, 255]]. Due to thread offset is not permuted, so the range of it are [[0, 0], [0, 63], [0,0], [0, 63], [0, 1]]. You can see block offset and thread offset is not aligned. So we should bitcast thread offsets to what we really want.

Reply:
We use [128, 256, 512]->[128, 512, 256] as example. Due to we add vectorized dim as the last dim of shape, so the input and output shape of transpose are [128, 256, 256, 2] - > [128, 256, 2, 256]. If block is permuted, the range of block offset are [[0, 127], [0, 255], [0, 1], [0, 255]]. Due to thread offset is not permuted, so the range of it are [[0, 0], [0, 63], [0,0], [0, 63], [0, 1]]. You can see block offset and thread offset is not aligned. So we should bitcast thread offsets to what we really want.

Reply:
Hmm, but why move the vector dim in the first place?

Reply:
Yes, of course it's caused by the permutation. My question as more about why the permutation set up to move the vector dimension in the first place. I don't really see the benefit, and it seems to make things a lot more complicated?

Reply:
I can not get correct block offsets without permutation (or you have another idea?). Block offsets [a, b, c, vec_dim] corresponds to [a, c, vec_dim, b].

Reply:
Just to make sure we're talking about the same thing, let's look at a toy example with 4x4 blocks:

If we have a 4x4 block starting at [a, b, c] (ignoring the vector dimension because its offset is always 0/size always 1):

```
0 1 2 3
4 5 6 7
8 9 a b
c d e f
```

With vectorization, thread 0 will read 0 and 1 from global memory and write them to shared memory, and it will read 0 and 4 from shared memory and write it to the indices corresponding to 0 and 1 in block `[a, c, b]`. So I believe for the block offsets what you want is `[a, c, b, vec_dim]`, where `vec_dim` is irrelevant (again, because it's always 0).

Reply:
We can use more simple example to show what is the correct block offset:
x0 y0 z0 w0
x1 y1 z1 w1
x2 y2 z2 w2
Assume a = b = c = 0, x0 y0 belong to block0, z0 w0 belong to block1, x1 y1 belong to block2, z1 w1 belong to block3. The per transpose offset of block3 is [0, 1, 1, 0]. If [a, c, b, vec_dim] is correct, the post transpose offset of block3 is [0, 1, 1, 0]. It's obvious that the index of z1 is same as block3 offset. Assume a in range [0, 0], b in range [0, 2], c in range [0, 1], vec_dim in range [0, 1], so the linear index of z1 is 1 * 3 * 2 + 2 = 8. But the linear index of z1 should be 7 as the below example show:
x0 x1 x2
y0 y1 y2
z0 z1 z2
w0 w1 w2

Reply:
I'm sorry, but I don't follow these explanations. Let me try to implement this myself, maybe I'll see the problem then.

Reply:
You can see what I say above. I use your conclusion, but get wrong block offset result. https://github.com/openxla/xla/pull/12552#discussion_r1603577281

Reply:
https://github.com/openxla/xla/pull/12620 should work? Maybe I'm still missing something.

Reply:
But I don't really like that either. The block offset logic is really weird. Hmm.

Reply:
Now I understand what you mean. If has this line https://github.com/openxla/xla/pull/12620/files#diff-0210cd152294b1475469517ffba6b5ee578acc4e786b0dceeee6854aba445a29R148, the block offset is correct. But I think your PR do the same thing as me, so your main problem now is to make the logic simpler and easier to understand, right?

Reply:
I think mine is already a bit simpler, WDYT. I won't have time to look into replacing Tiling today, but once we do that, it'll be good, I believe. If you agree, would you mind merging the two PRs? Yours has the vector size logic and updated tests.

Reply:
Yeah, It seems your logic only change block offset computation but not thread offset computation.  You can merge them If your PR does the rest. Or you want me to merge ?

Reply:
Gotcha, I changed the other implementation to always use 128 threads as well. The `Tiling` abstraction was the main source of complexity here, so I removed that. I took your tests and most of your heuristic, but added a guard against exceeding the shmem budget and enabled vectorization for block sizes not divisible by 64 (unless there was a good reason not to do that?) Still need to run this against a larger test suite, but feel free to take a look already (still https://github.com/openxla/xla/pull/12620).

Reply:
#12620 seems fine. I have no other advice. I also wanted to abandon tiling at first, but I am not sure if openxla will accept this seemingly big change. Now it seems like I can be a little bolder in my subsequent PRs.  Thank you for your patient reply.

Reply:
The last thing: Maybe we can enable vectorization only for <32 bit datatype. I observed performance regression for some f32 shape.